realm/realm-library/src/main/java/io/realm/RealmList.java

```diff
@@ -1288,15 +1292,1274 @@ public void add(@Nullable E e) {
+                a = (T[]) Array.newInstance(componentType, size);
+            }
+
+            for (int i = 0; i < size; i++) {
+                final E e = osListOperator.get(i);
+                if (e == null || e instanceof Float) {
+                    //noinspection unchecked
+                    a[i] = (T) e;
+                } else {
+                    if (!(e instanceof Number)) {
+                        throw new ArrayStoreException(String.format(Locale.ENGLISH,
+                                "The element of type %1$s can't be stored into the array of %2$s",
+                                e.getClass().getName(), a.getClass().getComponentType()));
+                    }
+                    //noinspection unchecked
+                    a[i] = (T) Float.valueOf(((Number) e).floatValue());
+                }
+            }
+            return a;
+        }
+
+        if (componentType == Boolean.class) {
+            if (a.length < size) {
+                //noinspection unchecked
+                a = (T[]) Array.newInstance(componentType, size);
+            }
+
+            for (int i = 0; i < size; i++) {
+                final E e = osListOperator.get(i);
+                //noinspection unchecked
+                a[i] = (T) e;
+            }
+            return a;
+        }
+
+        if (componentType == Date.class) {
+            if (a.length < size) {
+                //noinspection unchecked
+                a = (T[]) Array.newInstance(componentType, size);
+            }
+
+            for (int i = 0; i < size; i++) {
+                final E e = osListOperator.get(i);
+                //noinspection unchecked
+                a[i] = (T) e;
+            }
+            return a;
+        }
+
+        if (componentType == byte[].class) {
+            if (a.length < size) {
+                //noinspection unchecked
+                a = (T[]) Array.newInstance(componentType, size);
+            }
+
+            for (int i = 0; i < size; i++) {
+                final E e = osListOperator.get(i);
+                //noinspection unchecked
+                a[i] = (T) e;
+            }
+            return a;
+        }
+
         return super.toArray(a);
     }
+
+    /**
+     * Returns newly allocated {@code long} array that contains all values in this {@link RealmList}.
+     * <p>
+     * This method is available only when the list is managed and the element type is implementing {@link Number} interface.
+     * <p>
+     * If this list contains {@code null} value, {@code valueForNull} is used instead.
+     *
+     * @return newly allocated {@code long} array.
+     * @throws IllegalStateException if this {@link RealmList} is not managed or element type is not implementing {@link Number}.
+     * @see #toLongArray()
+     * @see #toArray()
+     * @see #toArray(Object[])
+     */
+    @SuppressWarnings("WeakerAccess")
+    public long[] toLongArray(@SuppressWarnings("SameParameterValue") long valueForNull) {
```

Comment:
Why add these methods. It seems a lot of code for very little benefit compared to the standard `T toArray()` and adds quite a lot of code. ?

Reply:
I agree ... seems not quite useful + need to write lots of code.

Reply:
`T toArray()` and `T[] toArray(T[])` cannot return primitive array.
it means that users always need to wrapper array to get array of elements even if the list is not nullable.
That is huge overhead and no way for users to avoid that.

Reply:
Sure, but the collection API's in general have that problem and almost no one are complaining about thoses. Besides, these add quite a lot of corner cases where they fail as well, e..g calling `toDateArray()` on a boolean list or nullable values.

E.g. even Android only have special types for maps like `SparseIntArray`, but nothing for lists. 

Also, the implementation is just doing the `Long -> long` conversion internally so you are not really saving anything on the performance when reading the array out.

Reply:
Also, please notice `ArrayList<Long>.toArray()` won't return a primitive array either. So I think it is acceptable and there are common ways to do the converting which is not much slower that this.

Reply:
Hmm, OK. I'll remove these methods.

Reply:
removed